test/pylib/report_plugin.py

```diff
@@ -13,3 +18,20 @@ def pytest_runtest_makereport(self):
         outcome = yield
         report = outcome.get_result()
         report.nodeid = f"{report.nodeid}.{self.mode}.{self.run_id}"
+
+    @pytest.fixture(scope="function", autouse=True)
+    def allure_set_mode(self, request, mode):
+        try:
```

Comment:
A request: can we remove the autouse fixture if allure is not found?
As it stands, if there are 1000 Alternator tests, this "import allure" will be attempted 1000 times, and ugly log messages will appear 1000 times (!). There is no real need to do this.
Maybe you can use a `@cache` function to check if allure exists, and this function will only check it once (and log it once).

Reply:
Please also run `test/alternator/run` and see the run length it reports. I want this script to continue to work, and don't want to to be noticibly slowed down by this auto-use fixture, and don't want it to spew 1000 warnings to the output.

Reply:
Added a new fixture with scope session to check that allure is installed. This will reduce the amount of try to import allure, but still, it will be executed for each pytest call. With this amount, there should be no noticeable impact on performance.

Reply:
@xtrey I can see that you still `import allure` in the third line of the file. It looks like it should be removed, no?

Reply:
Ah, I see that you implemented the `is_allure_installed` check by trying to import, but notice that you need to use `allure` in the `allure_set_mode` function. Perhaps you should also do `import allure` inside `def allure_set_mode`, inside `if is_allure_installed`? AFAIK python caches imported modules, so if importing a module succeeds then calling `import allure` for the second time should be cheap.

Besides, what about the cache annotation that Nadav suggested? Can we use it here to cache the result?

Reply:
@piodul you are right. Modified code to use lazy import caching.
Python does cache imports, and in case when we have the allure installed it will have no impact on performance. But I suppose the problem will be when you do not have the module in your system, python will try to import allure with every test case since the module will not be in sys.modules.